src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemEventArgs.cs

```diff
@@ -17,14 +17,18 @@ public class FileSystemEventArgs : EventArgs
         /// </devdoc>
         public FileSystemEventArgs(WatcherChangeTypes changeType, string directory, string? name)
         {
+            ArgumentNullException.ThrowIfNull(directory);
+
             _changeType = changeType;
             _name = name;
-            _fullPath = Path.Join(Path.GetFullPath(directory), name);
 
-            if (string.IsNullOrWhiteSpace(name))
+            _fullPath = Path.GetFullPath((string.IsNullOrWhiteSpace(directory), string.IsNullOrWhiteSpace(name)) switch
             {
-                _fullPath = PathInternal.EnsureTrailingSeparator(_fullPath);
-            }
+                (true, true) => "." + PathInternal.DirectorySeparatorCharAsString,
+                (true, false) => name!,
+                (false, true) => PathInternal.EnsureTrailingSeparator(directory),
+                (false, false) => Path.Join(directory, name),
+            });
```

Comment:
Are there any drawbacks to using a switch expression here?

Reply:
generated assembly code might be less optimal, but I don't believe that this method is ever going to be on a hot path so we should rather not worry about that.